File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        if not root: return 0
        
        que = [[root, float(-inf)]]
        ans = 0
        
        while que:
            [curNode, maxVal] = que.pop(0)
            
            if curNode.val >= maxVal:
                maxVal = curNode.val
                ans += 1
                
            if curNode.left:
                que.append([curNode.left, maxVal])
            if curNode.right:
                que.append([curNode.right, maxVal])
        
        return ans

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var goodNodes = function(root) {
    //dfs
    let cnt = 0;
    // helper(root.left, root.val);
    // helper(root.right, root.val);
    helper(root, root.val)
    return cnt;
    
    function helper(node, maxVal){
        if(!node) return;
        
        if(node.val >= maxVal) cnt++;
        
        helper(node.left, Math.max(node.val, maxVal));
        helper(node.right, Math.max(node.val, maxVal))
    }
    
};